internal/testutil/acc/config_formatter.go

```diff
@@ -50,3 +57,184 @@ func sortStringMapKeys(m map[string]string) []string {
+		"cluster_type":   clusterTypeStr,
+		"name":           clusterName,
+		"backup_enabled": req.CloudBackup,
+	}
+	if req.DiskSizeGb != 0 {
+		clusterRootAttributes["disk_size_gb"] = req.DiskSizeGb
+	}
+	addPrimitiveAttributes(cluster, clusterRootAttributes)
+	cluster.AppendNewline()
+	for i, spec := range specs {
+		err = writeReplicationSpec(cluster, spec)
+		if err != nil {
+			return "", "", fmt.Errorf("error writing hcl for replication spec %d: %w", i, err)
+		}
+	}
+	cluster.AppendNewline()
+	if req.ResourceDependencyName != "" {
+		if !strings.Contains(req.ResourceDependencyName, ".") {
+			return "", "", fmt.Errorf("req.ResourceDependencyName must have a '.'")
+		}
+		err = setAttributeHcl(cluster, fmt.Sprintf("depends_on = [%s]", req.ResourceDependencyName))
+		if err != nil {
+			return "", "", err
+		}
+	}
+	return "\n" + string(f.Bytes()), clusterName, err
+}
+
+func writeReplicationSpec(cluster *hclwrite.Body, spec admin.ReplicationSpec) error {
+	replicationBlock := cluster.AppendNewBlock("replication_specs", nil).Body()
+	err := addPrimitiveAttributesViaJSON(replicationBlock, spec)
+	if err != nil {
+		return err
+	}
+	for _, rc := range spec.GetRegionConfigs() {
+		if rc.Priority == nil {
+			rc.SetPriority(7)
+		}
+		replicationBlock.AppendNewline()
+		rcBlock := replicationBlock.AppendNewBlock("region_configs", nil).Body()
+		err = addPrimitiveAttributesViaJSON(rcBlock, rc)
+		if err != nil {
+			return err
+		}
+		autoScalingBlock := rcBlock.AppendNewBlock("auto_scaling", nil).Body()
+		if rc.AutoScaling == nil {
+			autoScalingBlock.SetAttributeValue("disk_gb_enabled", cty.BoolVal(false))
+		} else {
+			autoScaling := rc.GetAutoScaling()
+			return fmt.Errorf("auto_scaling on replication spec is not supportd yet %v", autoScaling)
+		}
+		nodeSpec := rc.GetElectableSpecs()
+		nodeSpecBlock := rcBlock.AppendNewBlock("electable_specs", nil).Body()
+		err = addPrimitiveAttributesViaJSON(nodeSpecBlock, nodeSpec)
+	}
+	return err
+}
+
+// Helper function for adding "primitive" bool/string/int/float attributes of a struct.
+func addPrimitiveAttributesViaJSON(b *hclwrite.Body, obj any) error {
+	var objMap map[string]any
+	inrec, err := json.Marshal(obj)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(inrec, &objMap)
+	if err != nil {
+		return err
+	}
+	addPrimitiveAttributes(b, objMap)
+	return nil
+}
+
+func addPrimitiveAttributes(b *hclwrite.Body, values map[string]any) {
+	for _, keyCamel := range sortStringMapKeysAny(values) {
+		key := ToSnakeCase(keyCamel)
+		value := values[keyCamel]
+		switch value := value.(type) {
+		case bool:
+			b.SetAttributeValue(key, cty.BoolVal(value))
+		case string:
+			if value == "" {
+				continue
+			}
+			b.SetAttributeValue(key, cty.StringVal(value))
+		case int:
+			if value == 0 {
+				continue
+			}
+			b.SetAttributeValue(key, cty.NumberIntVal(int64(value)))
+		// int gets parsed as float64 for json
+		case float64:
+			b.SetAttributeValue(key, cty.NumberIntVal(int64(value)))
```

Comment:
don't we want to check for 0?

Reply:
Yes, see comment above https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2404#discussion_r1672131749